target/linux/generic/files/drivers/platform/mikrotik/rb_nvmem.c

```diff
@@ -0,0 +1,244 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * NVMEM layout driver for MikroTik Routerboard hard config cells
+ *
+ * Copyright (C) 2024 Robert Marko <robimarko@gmail.com>
+ * Based on the sysfs hard config driver by Thibaut VARÈNE <hacks+kernel@slashdirt.org>
+ * Comments documenting the format carried over from routerboot.c
+ */
+
+#include <linux/etherdevice.h>
+#include <linux/mod_devicetable.h>
+#include <linux/mtd/mtd.h>
+#include <linux/nvmem-consumer.h>
+#include <linux/nvmem-provider.h>
+#include <linux/platform_device.h>
+
+#include "rb_hardconfig.h"
+
+#ifdef CONFIG_CPU_BIG_ENDIAN
+#define RB_MAGIC_HARD		"draH"
+#else
+#define RB_MAGIC_HARD		"Hard"
+#endif
+
+struct hard_config_header {
+	u8 magic[4];
+} __packed;
+
+/*
+ * Routerboot tag nodes are u32 values:
+ * - The low nibble is the tag identification number,
+ * - The high nibble is the tag payload length (node excluded) in bytes.
+ * Tag nodes are 32bit-aligned and in cpu-endian.
+ */
+#ifdef CONFIG_CPU_BIG_ENDIAN
+struct hard_config_tlv {
+	u16 len;
+	u16 tag;
+} __packed;
+#else
+struct hard_config_tlv {
+	u16 tag;
+	u16 len;
+} __packed;
+#endif
+
+static const char *rb_tlv_cell_name(u16 tag)
+{
+	switch (tag) {
+	case RB_ID_FLASH_INFO:
+		return "flash-info";
+	case RB_ID_MAC_ADDRESS_PACK:
+		return "base-mac-address";
+	case RB_ID_BOARD_PRODUCT_CODE:
+		return "board-product-code";
+	case RB_ID_BIOS_VERSION:
+		return "booter-version";
+	case RB_ID_SERIAL_NUMBER:
+		return "board-serial";
+	case RB_ID_MEMORY_SIZE:
+		return "mem-size";
+	case RB_ID_MAC_ADDRESS_COUNT:
+		return "mac-count";
+	case RB_ID_HW_OPTIONS:
+		return "hw-options";
+	case RB_ID_WLAN_DATA:
+		return "wlan-data";
```

Comment:
No, NVMEM cells are not constantly in memory, the layout driver just finds them and populates the cell information, so then they are read directly from the underlying storage only when requested by code, so there is no memory being used otherwise.